### ISSUE TITLE
Rucio implementation of Storage stage-out and stage-in operations

### ISSUE THREAD

**Issue by nsmith-** [dmwm/WMCore#9622]
**Impact of the new feature**
This impacts the job wrapper and partially the WMAgent data injector routine.

**Is your feature request related to a problem? Please describe.**
No problems yet, planning for the future

**Describe the solution you'd like**
Rucio provides two encapsulated methods to stage in and out via the CLI: `rucio download` and `rucio upload`. The job wrapper should use these methods to stage in and out files when requested to via the 'rucio' implementation. A necessary parameter would be the RSE name.

**Describe alternatives you've considered**
The alternative would be to poll the LFN-to-PFN algo for the site as done in PhEDEx implementation.

**Additional context**
This is motivated by two pending changes:

1. If this is implemented, it would likely be very easy to set the stageout RSE to a "test" RSE, as discussed in https://github.com/dmwm/WMCore/pull/9617/files/bbd1bfd0ddc890ed8a848bfb0971461f17f52135#r406247530 which would allow us to do test workflows using real sites.
2. There is a planned change to the site storage specification, migrating from storage.xml to a new data source. Letting rucio handle the operation removes the need for WMCore to adapt to that change.

**Comment by ericvaandering** [comment]
If I understand what you're proposing, this is a big change since we don't download files to the worker node as ATLAS does. We've discussed this in the past, but without a consensus.

**Comment by nsmith-** [comment]
No, that's not exactly what I'm saying, I don't think we'll be *using* stage-in but since the abstract storage provides both as far as I can tell from looking at the code here, I'm just saying one should implement both.

**Comment by nsmith-** [comment]
I'm really only interested in the stage-out part

**Comment by amaltaro** [comment]
As from the quick discussion on this in yesterday's Rucio Transition meeting. We have two options:
*) adopt rucio-upload (and I guess rucio-download) for stage out/stage in, deprecating all our Storage/Backend plugins
*) refactor how LFNxPFNs, protocols, etc are handled; moving from storage.xml to storage.json

**Comment by amaltaro** [comment]
The storage.json migration was brought up in today's O&C weekly meeting.

@ericvaandering @nsmith- can you please remind me again whether the adoption of rucio-upload / rucio-download would be an alternative for this storage.xml/storage.json . Would you see any other use case where we would still rely on these storage.xml/json files?

**Comment by ericvaandering** [comment]
Rucio upload and especially download would be a complete change of the CMS model requiring much more scratch space on the worker nodes. I didn't think we were even considering this.

Add to that that I don't know if rucio upload and download have really been tested for CMS let alone at scale and I'm confused why we're talking about this vs. a simple change of parsing.

**Comment by nsmith-** [comment]
That's not exactly true Eric. As the WM certificate has permission to write directly to the destination, it can use `rucio upload` without need for a temporary RSE or scratch space. What I want to leverage is rucio LFN-to-PFN translation rather than having a standalone file.

**Comment by nsmith-** [comment]
As for download, I agree we don't really have any use case for it in production. But, it may be preferable to use rucio to locate files (as ATLAS does) in a future where we have more cache systems.

**Comment by ericvaandering** [comment]
Yes, it should, assuming you know the name of the RSE for the system you land on. Stefano is using this API for CRAB in some capacity too.

**Comment by amaltaro** [comment]
Going through this discussion once again, I think the best decision here would be to avoid communication to the Rucio server from the worker nodes. Instead, we could make all the necessary translations on the agent side and transfer a (JSON) file to the worker node, once a job gets matched to a pilot.

For that to happen, the agent would have to create two mappings:
1) map of PSN to RSE (old PNN). So a map of processing site name to the storage element. Here I assume that a site can only be mapped to one Disk RSE. Please let me know if this is wrong!
2) a map of LFN to PFN, for all RSEs (and for all protocols?)

With that, we would have the RSE name and be able to lookup what PFN to use. But the real benefit in that would come from the use of the `rucio upload` command to simplify all the stage out plugins and protocols that we use in WMCore.

Problem is, AFAIK `rucio upload` performs two actions:
1) it registers a file into the Rucio server
2) if successful, it then tries to upload that file to the target endpoint.

I think the best here would be to have a mechanism to skip the file injection into Rucio. Such that we simply use it to stage out files to the storage, leaving the file registration up to the RucioInjector component (as it is currently implemented).

@ericvaandering would you see any problem in the proposal above? I guess we would have to request a new feature for Rucio to allow file upload-only, right?
